src/python/pants/backend/codegen/export_codegen_goal.py

```diff
@@ -46,26 +46,25 @@ async def export_codegen(
     # We run all possible code generators. Running codegen requires specifying the expected
     # output_type, so we must inspect what is possible to generate.
     all_generate_request_types = union_membership.get(GenerateSourcesRequest)
-    inputs_to_outputs = {
-        req.input: req.output for req in all_generate_request_types if req.exportable
-    }
+    inputs_to_outputs = [
+        (req.input, req.output) for req in all_generate_request_types if req.exportable
+    ]
     codegen_sources_fields_with_output = []
     for tgt in targets:
         if not tgt.has_field(SourcesField):
             continue
         sources = tgt[SourcesField]
-        for input_type in inputs_to_outputs:
+        for input_type, output_type in inputs_to_outputs:
             if isinstance(sources, input_type):
-                output_type = inputs_to_outputs[input_type]
                 codegen_sources_fields_with_output.append((sources, output_type))
 
     if not codegen_sources_fields_with_output:
         codegen_targets = sorted(
             {
                 tgt_type.alias
                 for tgt_type in registered_target_types.types
-                for input_sources in inputs_to_outputs.keys()
-                if tgt_type.class_has_field(input_sources, union_membership=union_membership)
+                for input_source, _ in inputs_to_outputs
```

Comment:
not a super big thing, as there's likely few entries and even fewer duplicated entries, but still, the `Target.class_has_field` is not a simple lookup, so would prefer to eliminate redundant calls to it.

```suggestion
                for input_source in {input_source for input_source, _ in inputs_to_outputs}
```

Reply:
Ah, of course. Thanks!